ffi/src/scan.rs

```diff
@@ -0,0 +1,392 @@
+    // Also keep a reference to the external client for its error allocator.
+    // Parquet and Json handlers don't hold any reference to the tokio reactor, so the iterator
+    // terminates early if the last table client goes out of scope.
+    engine_interface: Arc<dyn ExternEngineInterface>,
+}
+
+impl BoxHandle for KernelScanDataIterator {}
+
+impl Drop for KernelScanDataIterator {
+    fn drop(&mut self) {
+        debug!("dropping KernelScanDataIterator");
+    }
+}
+
+/// Get an iterator over the data needed to perform a scan. This will return a
+/// [`KernelScanDataIterator`] which can be passed to [`kernel_scan_data_next`] to get the actual
+/// data in the iterator.
+///
+/// # Safety
+///
+/// Engine is responsible for passing a valid [`ExternEngineInterfaceHandle`] and [`Scan`]
+#[no_mangle]
+pub unsafe extern "C" fn kernel_scan_data_init(
+    engine_interface: *const ExternEngineInterfaceHandle,
+    scan: *mut Scan,
+) -> ExternResult<*mut KernelScanDataIterator> {
+    kernel_scan_data_init_impl(engine_interface, scan).into_extern_result(engine_interface)
+}
+
+unsafe fn kernel_scan_data_init_impl(
+    engine_interface: *const ExternEngineInterfaceHandle,
+    scan: *mut Scan,
+) -> DeltaResult<*mut KernelScanDataIterator> {
+    let engine_interface = unsafe { ArcHandle::clone_as_arc(engine_interface) };
+    let boxed_scan = unsafe { Box::from_raw(scan) };
+    let scan = boxed_scan.kernel_scan;
+    let scan_data = scan.scan_data(engine_interface.table_client().as_ref())?;
+    let data = KernelScanDataIterator {
+        data: Box::new(scan_data),
+        engine_interface,
+    };
+    Ok(data.into_handle())
+}
+
+/// # Safety
+///
+/// The iterator must be valid (returned by [kernel_scan_data_init]) and not yet freed by
+/// [kernel_scan_data_free]. The visitor function pointer must be non-null.
+#[no_mangle]
+pub unsafe extern "C" fn kernel_scan_data_next(
+    data: &mut KernelScanDataIterator,
+    engine_context: *mut c_void,
+    engine_visitor: extern "C" fn(
+        engine_context: *mut c_void,
+        engine_data: *mut EngineDataHandle,
+        selection_vector: KernelBoolSlice,
+    ),
+) -> ExternResult<bool> {
+    kernel_scan_data_next_impl(data, engine_context, engine_visitor)
+        .into_extern_result(data.engine_interface.error_allocator())
+}
+fn kernel_scan_data_next_impl(
+    data: &mut KernelScanDataIterator,
+    engine_context: *mut c_void,
+    engine_visitor: extern "C" fn(
+        engine_context: *mut c_void,
+        engine_data: *mut EngineDataHandle,
+        selection_vector: KernelBoolSlice,
+    ),
+) -> DeltaResult<bool> {
+    if let Some((data, sel_vec)) = data.data.next().transpose()? {
+        let bool_slice: KernelBoolSlice = sel_vec.into();
+        let data_handle = BoxHandle::into_handle(EngineDataHandle { data });
+        (engine_visitor)(engine_context, data_handle, bool_slice);
+        // ensure we free the data
+        unsafe { BoxHandle::drop_handle(data_handle) };
```

Comment:
We need the box handle because `EngineData` is a trait. So if you try and have the callback just take a `*const dyn EngineData` you'll get warned that it's not ffi safe because it's a trait. So I think we have to wrap it into a non-trait and then pass that. Potentially we could use an `Arc` rather than a box

Reply:
Circling back around: This PR seems to have exposed a big unaddressed question, of how to best handle the various types that we need to pass across the FFI boundary: sized vs unsized, local vs. other-crate, owned vs. borrowed, arc vs. box, etc. The current set of tools appears to be insufficient and I'm still chewing on ideas for how to improve the situation.

Reply:
Note that _even_ if we got all this working, we _still_ can't write something like:
```rust
extern "C" fn use_data(foo: Box<dyn EngineData>) { ... }
```

you'll get:
```
warning: `extern` fn uses type `Box<dyn EngineData>`, which is not FFI-safe
   --> [file].rs:367:29
    |
367 | extern "C" fn use_data(foo: Box<dyn EngineData>) {  }
    |                             ^^^^^^^^^^^^^^^^^^^ not FFI-safe
    |
    = note: box cannot be represented as a single pointer
    = note: `#[warn(improper_ctypes_definitions)]` on by default
```

I have not yet dug into why `Box<Foo>` works, but `Box<dyn EngineData>` does not.

Reply:
Yeah, I had thought perhaps the various `Rc::X_ptr` functions might have something useful, but indeed they will all return a fat pointer if the Rc is holding a trait object.

I did come across [`thin`](https://docs.rs/thin/0.2.0/thin/) which does some of what we want. I need to think more about the semantics when the _Engine_ does the allocation, and how we would make that work, but it could get us what we want.

Otherwise it seems a double box of some sort will be needed.